bst.rb

```diff
@@ -0,0 +1,322 @@
+  end
+
+  def remove(value, node = self.root)
+    rh(value, node = self.root)
+    @size -=1
+    node
+  end
+
+  private
+  def rh(value, node = self.root)
+    return nil if node == nil
+    if node.value > value
+      node.left = rh(value, node.left)
+    elsif node.value < value
+      node.right = rh(value, node.right)
+    else
+      if node.left != nil && node.right != nil
+        temp = node
+        rmin = find_min(node.right)
+        node.value = rmin.value
+        node.right = rh(rmin.value, node.right)
+      elsif node.left != nil
+        node = node.left
+      elsif node.right != nil
+        node = node.right
+      else
+        node = nil
+      end
+    end
+    node
+  end
+end
+
+t = BST.new()
+
+while true
+  puts "Select from the given options"
+  puts "  Enter 0: to insert multiple elements"
+  puts "  Enter 1: to insert value"
+  puts "  Enter 2: to delete value"
+  puts "  Enter 3: to print BST"
+  puts "  Enter 4: to find Max"
+  puts "  Enter 5: to find Min"
+  puts "  Enter 6: to find if value exists"
+  puts "  Enter 7: to print all paths from root"
+  puts "  Enter 8: to load a new bst from the file"
+  puts "  Enter quit to save and exit"
+  puts "  Enter anything else to exit"
+
+  option = gets.chomp
+  system("clear")
+  if !option.is_i?
+    t.save() if option.downcase == "quit"
+    break
+  end
+  option = option.to_i
+
+  if option == 0
+    puts "enter multiple elements sepreated with ,"
+    arr = gets.chomp.split(",").map(&:to_i)
+    arr.each{|i| t.insert(i)}
+    puts "\n\n"
+    next
+  end
+
+  if option == 1
+    puts "Enter the value you want to insert"
+    v = gets.to_i
+    t.insert(v)
+    puts "\n#{v} was inserted in the tree"
+    puts "size of the tree now is #{t.size} \n\n"
+    next
+  end
+
+  if option == 2
+    puts "Enter the value you want to delete"
+    v = gets.to_i
+    if t.contains?(v)
+      t.remove(v)
+      puts "\n#{v} was deleted from the tree"
+      puts "size of the tree now is #{t.size} \n\n"
+    else
+      puts "value not found can't delete it \n\n"
+    end
+    next
+  end
+
+  if option == 3
+    puts "Printing the tree in InOrder"
+    t.print_inorder()
+    puts "end\n Printing the tree in PreOrder"
+    t.print_preorder()
+    puts "end\n Printing the tree in PostOrder"
+    t.print_postorder()
+    puts "end\n Printing the tree in LevelOrder"
+    t.print_levelorder()
+    puts "end\n\n"
+    next
+  end
+
+  if option == 4
+    puts "the maximum is #{t.find_max().value}\n\n"
+    next
+  end
+
+  if option == 5
+    puts "the minimum is #{t.find_min().value}\n\n"
+    next
+  end
+
+  if option == 6
+    puts "Enter value to search it in treee"
+    v = t.contains(gets.to_i)
+    puts v ? "value exists" : "value does't exists"
+    next
+  end
+
+  if option == 7
+    puts "Printing all paths from Root"
+    t.pathp()
+    puts "\n\n"
+    next
+  end
+
+  if option == 8
+    t.loadit()
+    puts "Data loaded from the file\n\n"
+  end
+end
```

Comment:
logic is good
followed coding guidelines on 99% places
function names poor
no comments in the code